### PR TITLE
specify elements which should continue to receive touchmove events

### PR DESCRIPTION
**Pull request by diachedelic**
Workaround for #40 , make `disableScrollLock` ignore any element with the `body-scroll-lock-ignore` attribute set (or whose ancestor has it set).

### CODE REVIEW

**Comment by diachedelic** [comment]
By ignore, I mean it does not interfere with `touchmove` events.

**Comment by LarsDenBakker** [comment]
Walking the DOM is expensive. If this needs to be supported, make it an option at least.

**Comment by diachedelic** [comment]
If it must be an option, perhaps it could be a user defined function, e.g:

```
disableBodyScroll(container, {
  ignoreEl: (el) => { return el.tagName === 'TEXTAREA' }
}
```

That way optimising the performance is up to the user, who may not need to do any DOM walking.

**Comment by diachedelic** [comment]
I've implemented an `allowTouchMove` option, to avoid forcing a DOM walk.

## Simple case
```javascript
disableBodyScroll(container, {
  allowTouchMove: el => (el.tagName === 'TEXTAREA')
}
```

## more complex case 
Ignores any element which is, or is a child of, an element which has the `body-scroll-lock-disable` attribute
```javascript
disableBodyScroll(container, {
  allowTouchMove: el => {
    while (el && el !== document.body) {
      if (el.getAttribute('body-scroll-lock-ignore') !== null) {
        return true
      }

      el = el.parentNode
    }
  },
})
```

```html
<div id="container>
  <div id="scrolling-map" body-scroll-lock-disable>
    ...
  </div>
</div>
```

**Comment by willmcpo** [comment]
Thanks for the PR. I will review this and other PRs. I'm just going to be busy for another 2 weeks, so bear with me :).

**Comment by diachedelic** [comment]
Thanks!

I can also add something to the readme about this option, once I know it's going to be approved.

File: src/bodyScrollLock.js [hunk 1]
```diff
@@ -14,14 +20,29 @@ const isIosDevice =
...
+      return true;
+    }
+
+    return false;
+  });
+
 const preventDefault = (rawEvent: HandleScrollEvent): boolean => {
   const e = rawEvent || window.event;
+
+  if (allowTouchMove(e.target)) {
```
[the remark this answers is not in the document]
Comment (diachedelic) [reply src/bodyScrollLock.js]: Hmm i remember this being necessary,  but maybe I forgot to try calling `event.stopPropagation();` after line 103.  I will test that and get back to you.

**Comment by willmcpo** [comment]
Can you provide an example of what the supplief allowTouchMove logic could look like?

It looks like the consumer will provide that logic through the BodyScrollOption, so the consumer has control over which elements should bypass the BSL logic right?

**Comment by diachedelic** [comment]
> Can you provide an example of what the supplied allowTouchMove logic could look like?

Is [this what you're after](https://github.com/willmcpo/body-scroll-lock/pull/41#issuecomment-424572597)?

> It looks like the consumer will provide that logic through the BodyScrollOption, so the consumer has control over which elements should bypass the BSL logic right?

Yes

> At line 142, why not just look for the allowTouchMove option right there, and if there, call it to determine whether to call handleScroll or not?

> This will avoid what seems like unnecessary looping through the locks array.

Yes thats a good point.

File: src/bodyScrollLock.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (diachedelic) [reply src/bodyScrollLock.js]: Okay - for things like textareas and range sliders, stopping propagation in `handleScroll` is enough (don't need the above check in `preventDefault`).  

However, Leaflet actually adds a `touchmove` event listener to `document`, so calling `event.stopPropagation()` in `handleScroll` prevents the map panning at all.

All I can think of is to leave the checks as they are, or let `allowTouchMove` return `true`, `false`, or `"propagate"` (indicating the events should be allowed to propagate up to `document`)

[the remark this answers is not in the document]
Comment (diachedelic) [reply src/bodyScrollLock.js]: or `"bubble"` might be a bit better

**Comment by aaronstezycki** [comment]
There any progress on this pull? Looks like a nice addition enabling multiple elements to continue to scroll once the body has been disabled.

File: src/bodyScrollLock.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (diachedelic) [reply src/bodyScrollLock.js]: What do you think?

**Comment by diachedelic** [comment]
Looks like we got stuck figuring out the implementation details...

**Comment by willpox** [comment]
Hi guys - sorry it is on my to-do list, I'm aiming to have this one merged in by next week some time.

**Comment by diachedelic** [comment]
nice one @willpox !  Let me know if there's anything I can do

File: src/bodyScrollLock.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (willpox) [reply src/bodyScrollLock.js]: Ok

**Comment by willmcpo** [comment]
Published npm version 2.5.11-beta.1

Let me know if anyone has any issues with it

**Comment by diachedelic** [comment]
`2.5.11-beta.1` is tested and working on my iOS simulator

**Comment by pmochine** [comment]
@diachedelic thank you! it worked :)

**Pull request merged**